fs/nfsd/nfs4acl.c

```diff
@@ -751,26 +906,161 @@ static int nfs4_acl_nfsv4_to_posix(struct nfs4_acl *acl,
 	return ret;
 }
 
-__be32 nfsd4_acl_to_attr(enum nfs_ftype4 type, struct nfs4_acl *acl,
-			 struct nfsd_attrs *attr)
+static __be32
+nfsd4_acl_to_attr_posix(enum nfs_ftype4 type, struct nfs4_acl *acl,
+			fsacl_t *fsaclp)
 {
 	int host_error;
 	unsigned int flags = 0;
 
 	if (!acl)
 		return nfs_ok;
-
 	if (type == NF4DIR)
 		flags = NFS4_ACL_DIR;
 
-	host_error = nfs4_acl_nfsv4_to_posix(acl, &attr->na_pacl,
-					     &attr->na_dpacl, flags);
+	host_error = nfs4_acl_nfsv4_to_posix(acl, &fsaclp->posixacl.na_pacl,
+					     &fsaclp->posixacl.na_dpacl, flags);
 	if (host_error == -EINVAL)
 		return nfserr_attrnotsupp;
 	else
 		return nfserrno(host_error);
 }
 
+static int
+convert_ace_to_nfs41(u32 *p, const struct nfs4_ace *ace)
+{
+	int error = 0;
+	nfsace4i nace;
+
+	nace = (nfsace4i) {
+		.type = (acetype4)ace->type,
+		.flag = (aceflag4)ace->flag & NFS41_FLAGS,
+		.access_mask = ace->access_mask & NFS4_ACE_MASK_ALL,
+		.who = -1,
+	};
+
+	/* Audit and Alarm are not currently supported */
+	if (nace.type > NFS4_ACE_ACCESS_DENIED_ACE_TYPE)
+		return -EINVAL;
+
+	switch (ace->whotype) {
+	case NFS4_ACL_WHO_OWNER:
+		nace.iflag = ACEI4_SPECIAL_WHO;
+		nace.who = ACE4_SPECIAL_OWNER;
+		break;
+	case NFS4_ACL_WHO_GROUP:
+		nace.iflag = ACEI4_SPECIAL_WHO;
+		nace.who = ACE4_SPECIAL_GROUP;
+		break;
+	case NFS4_ACL_WHO_EVERYONE:
+		nace.iflag = ACEI4_SPECIAL_WHO;
+		nace.who = ACE4_SPECIAL_EVERYONE;
+		break;
+	case NFS4_ACL_WHO_NAMED:
+		if (ace->flag & NFS4_ACE_IDENTIFIER_GROUP)
+			nace.who = (int)__kgid_val(ace->who_gid);
+		else
+			nace.who = (int)__kuid_val(ace->who_uid);
+
+		BUG_ON(nace.who == -1);
+		break;
+	default:
+		error = -EINVAL;
+		break;
+	}
+
+	*p++ = htonl(nace.type);
+	*p++ = htonl(nace.flag);
+	*p++ = htonl(nace.iflag);
+	*p++ = htonl(nace.access_mask);
+	*p++ = htonl(nace.who);
+
+	return error;
+}
+
+static int
+generate_nfs41acl_buf(u32 *xdrbuf, const struct nfs4_acl *acl)
+{
+	int error = 0;
+	int i;
+
+	/* first byte is NFS41 Flags. Skip since these are RFC3530 acls */
+	*xdrbuf++ = 0;
+	*xdrbuf++ = htonl(acl->naces);
+
+	for (i = 0; i < acl->naces; i++, xdrbuf += NACE41_LEN) {
+		error = convert_ace_to_nfs41(xdrbuf, &acl->aces[i]);
+		if (error)
+			break;
+	}
+
+	return error;
+}
+
+static __be32
+nfsd4_acl_to_attr_zfsacl(enum nfs_ftype4 type, struct nfs4_acl *acl,
+			 fsacl_t *fsaclp)
+{
+	int error;
+	u32 *xdr_buf = NULL;
+	size_t len;
+
+	if (!acl)
+		return nfs_ok;
+
+	if (acl->naces > NFS41ACL_MAX_ENTRIES)
+		return nfserrno(-ERANGE);
+
+	else if (acl->naces == 0)
+		return nfserrno(-EINVAL);
+
+	len = ACES_TO_XDRSIZE(acl->naces);
+
+	xdr_buf = kzalloc(len, GFP_KERNEL);
+	if (!xdr_buf)
+		return nfserrno(-ENOMEM);
+
+	error = generate_nfs41acl_buf(xdr_buf, acl);
+	if (error) {
+		kfree(xdr_buf);
+		return nfserrno(error);
+		return error;
```

Comment:
Two return statements?  Presumably should just be
`		return nfserrno(error);
`